### ISSUE TITLE
None-square SVG files don't render correctly

### ISSUE THREAD

**Issue by ewaldc** [openhab/openhab-android#2021]
### Actual behaviour
Non-square-SVG files (e.g. viewBox="0 0 1200 1000") don't render properly (scaling is off, part of image is off screen, see picture). Square images render fine.

### Expected behaviour
Rendering should as in attached image (patched version of the code, see attached).
A user workaround is of course to make the SVG square.

### Steps to reproduce
Render attached SVG. 

### Can you reproduce the issue in demo mode?
Yes, by replacing chart.png with chart.svg for weather_chart

#### Client
* Android version: multiple (7.x, 8.x, 9.x, 10.x)
* Device model: at least 8 different phones and tablets
* App version : most recent stable and beta (2.12.20-beta)
* Build flavor : 
* Device language: english

#### Server
* Server version: 2.5.5
* Reverse Proxy: 
* Authentication method : 

I think I discovered where the issue is: in [ExtensionFuncs.kt](https://github.com/openhab/openhab-android/blob/master/mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt)

```
        if (docWidth < 0 || docHeight < 0) {
            val aspectRatio = svg.documentAspectRatio
            if (aspectRatio > 0) {
                val heightForAspect = targetSize.toFloat() / aspectRatio
                val widthForAspect = targetSize.toFloat() * aspectRatio
                if (widthForAspect < heightForAspect) {
                    docWidth = Math.round(widthForAspect)
                    docHeight = targetSize
                } else {
                    docWidth = targetSize
                    docHeight = Math.round(heightForAspect)
                }
            } else {
                docWidth = targetSize
                docHeight = targetSize
            }
            // we didn't take density into account anymore when calculating docWidth
            // and docHeight, so don't scale with it and just let the renderer
            // figure out the scaling
            density = null
        }
        if (docWidth != targetSize || docHeight != targetSize) {
            val scaleWidth = targetSize.toFloat() / docWidth
            val scaleHeight = targetSize.toFloat() / docHeight
            density = (scaleWidth + scaleHeight) / 2
        }

        val bitmap = Bitmap.createBitmap(targetSize, targetSize, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        if (density != null) {
            canvas.scale(density, density)
        }
```
It's fixed by changing:
`        if (docWidth != targetSize || docHeight != targetSize) {`
to
`        if (docWidth != targetSize && docHeight != targetSize) {`
Basically there is no need for _density_ since properly build SVG images will autoscale fine and can preserve aspect ratio (e.g. with _preserveAspectRatio="xMidYMid meet"_) but I admit I don't fully understand the whole purpose of _canvas.scale(density, density)_

![Screenshot_20200524-100919_openHAB_Beta](https://user-images.githubusercontent.com/14278196/82749766-c64b1800-9dab-11ea-8582-065faa52dd8a.png)
![Screenshot_20200524-102338_openHAB_Beta](https://user-images.githubusercontent.com/14278196/82749767-c6e3ae80-9dab-11ea-9e89-0d71557d5461.png)
[chart.zip](https://github.com/openhab/openhab-android/files/4673176/chart.zip)

**Comment by maniac103** [comment]
> but I admit I don't fully understand the whole purpose of canvas.scale(density, density)

The point of that scaling is scaling the document size to the view size when rendering the SVG to a bitmap (SVGs are not supported natively).

TBH, I don't understand the logic in that block either. This is the code (I wrote) this extension function originated from: https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/utils/HttpImageGetter.java#L560 ... as you can see, width and height are always scaled by the same factor there.

**Comment by ewaldc** [comment]
Wow, if that is not a real-time response...
The original code is not exactly the same if I read it well as it basically simply divides density, docWidth and docHeight by 2 as to make sure all rendering will be inside the canvas (kind of compensates for badly written SVG IMHO...) and that should work fine.  However in Habdroid it ended up with a scaling factor >1 (_density_) the way it's written using the attached SVG.  And of course, that will cause the rendering to go off-screen.  
_Original_
```
                while (docWidth >= maxWidth || docHeight >= maxHeight) {
                    docWidth /= 2;
                    docHeight /= 2;
                    if (density != null) {
                        density /= 2;
                    }
                }
```
_Habdroid_
```
        if (docWidth != targetSize || docHeight != targetSize) {
            val scaleWidth = targetSize.toFloat() / docWidth
            val scaleHeight = targetSize.toFloat() / docHeight
            density = (scaleWidth + scaleHeight) / 2
        }
```
Somehow I could not find a use case with a properly written SVG where that scaling is needed.
In addition, many SVG icon sets are square so it's not that straightforward to hit this issue...
I only hit it because I was replacing the Grafana charting library by SVG which requires almost no CPU resources on small devices (e.g. on RPI).

**Comment by ewaldc** [comment]
As I read it with this new knowledge:
SVG's with no height and width, but just a viewbox:
`        if (docWidth < 0 || docHeight < 0) {` 
will be true and a docWidth and docHeight is (correctly) calculated and density set to null.
In that case no scaling is needed, so the code sequence ` if (docWidth != targetSize || docHeight != targetSize) {` _should be skipped_ and density should remain _null_

For SVG's with height and/or width:
`        if (docWidth < 0 || docHeight < 0) {` 
will be false, and 
`        if (docWidth != targetSize || docHeight != targetSize) {`
should indeed be checked and is correctly calculated IMHO, resulting in <1 scaling factor.

So my proposed fix was wrong...
It should simple not execute  ` if (docWidth != targetSize || docHeight != targetSize) {` when there was no docWidth/docHeight.

**Comment by ewaldc** [comment]
Ah, there is another difference that could explain the problem:
Original code uses the new calculated values for docWidth and _docHeight_:
 `Bitmap bitmap = Bitmap.createBitmap(_docWidth_, _docHeight_, Bitmap.Config.ARGB_8888);`
Habdroid uses targetsize, which make the bitmap square, but at the same time ignores the whole _docWidth_/_docHeight_ calculation above...
`val bitmap = Bitmap.createBitmap(targetSize, targetSize, Bitmap.Config.ARGB_8888)`

Maybe there was a reason to change this e.g. make sure the image is always square ?

**Comment by maniac103** [comment]
Code was introduced in #1194, FWIW.

**Comment by mueller-ma** [comment]
> Maybe there was a reason to change this e.g. make sure the image is always square ?

The code was introduced by myself when I fixed a scaling issue (#728). At that time I didn't think at image items having SVG images. Some time later I realized that this might cause issues, but IMO this issue was mostly theoretically, because - at that time - no one used SVG images in image items (or charts, which use the same rendering code). So I didn't open an issue.


A bit off topic: 

> I only hit it because I was replacing the Grafana charting library by SVG which requires almost no CPU resources on small devices (e.g. on RPI).

Are you planing to contribute that to the server? I added a preference to mitigate the issue with slow loading charts by reducing the size of the rendered image: https://www.openhab.org/docs/apps/android.html#chart-loading-is-too-slow
But of course using SVG there is a much better fix.

> Habdroid 

It isn't called Habdroid anymore [since 2017](#294) ;)

**Comment by maniac103** [comment]
> The code was introduced by myself when I fixed a scaling issue (#728).

Do you remember whether there was a strong reason for forcing the resulting bitmap to be square? If not, I think we should just preserve aspect ratio when calculating density, that is, let the larger value of docWidth and docHeight match targetSize, adjust density to that scale and use targetSize * aspect ratio as other dimension.

**Comment by mueller-ma** [comment]
> Do you remember whether there was a strong reason for forcing the resulting bitmap to be square?

Well, I didn't think of non-square SVG images at that time.

> I think we should just preserve aspect ratio

Yes, we should. I started with that in #1893.

**Comment by ewaldc** [comment]
> Habdroid
Sorry, failing brain + laziness + glass of red wine.  What is an acceptable shorthand?    

> Do you remember whether there was a strong reason for forcing the resulting bitmap to be square? If not, I think we should just preserve aspect ratio when calculating density, that is, let the larger value of docWidth and docHeight match targetSize, adjust density to that scale and use targetSize * aspect ratio as other dimension.

Perfect.  I think that is what the original code does with maxWidth/Height instead of (one) targetSize
That is how I changed the code and it worked, but I wanted to create a number of SVG test cases just to 100% sure before submitting a PR.

**Comment by ewaldc** [comment]
@maniac103, @mueller-ma 
Thanks for your patience and efforts to fix this issue. You guys are wonderful!
Did not realize though it was that complex when I filed it... Will test ASAP.  If there is anything I can do, please let me know.
